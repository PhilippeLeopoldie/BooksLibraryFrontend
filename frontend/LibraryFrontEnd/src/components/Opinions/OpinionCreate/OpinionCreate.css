.opinionCreateCard--grid {
    display: grid;
    grid-template-rows: 20% 65% 15%;
    width: 18rem;
    min-width: 18rem;
    height: 29rem;
    margin: 0 1rem 2rem 1rem;
    border: 1px var(--cardbordercolor_light) solid;
    border-radius: 10px;
}

.opinionCreateCard--grid:hover {
    border: 1px var(--cardborder-hover) solid;
}

.opinionCreate__header {
  border-radius: 10px;
}

.opinionCreate__booktitle{
    max-height:3.5rem;
    overflow:auto;
}

.opinionCreate--natural {
    background: var(--backgroundcolor-whitesmoke);
}

.opinionCreate--black {
    background: var(--backgroundcolor-darkmode);
    color: var(--fontcolor-dark);
}

.opinionCreateCard--black--grid {
    border: 1px var(--cardbordercolor_dark) solid;
}

.opinionCreate__main--grid {
  display:grid;
  grid-template-rows: 25% 75%;
}
.opinionCreate__rateText {
  font-weight: bold;
  font-size: 1.125rem;
}

.opinionCreate__inputs--grid {
  display: grid;
  grid-template-rows: 65% 20% 15%;
}

.opinionCreate__view {
    margin: 0 2%;
    border-radius: 10px;
    background-color: inherit;
    color: inherit;
    font-size: 1.25rem;
    max-width: 16.5rem;
    resize: none;
    border: 1px var(--forminputcolor) solid;
}

.opinionCreate__input {
    margin: 2%;
    max-width: 16.5rem;
    font-size: 1.25rem;
    color: inherit;
    border: 1px var(--forminputcolor) solid;
}

.opinionCreate__footer--flex {
  display: flex;
  justify-content: space-between;
  margin-bottom: 20%;
}
.validation__errorMessage {
  color: red;
  background-color: transparent;
  font-size:1rem;
  font-weight: bold;
}

.opinionCreate__postButton {
  font-weight: bold;
  font-size: 1.125rem;
  margin: auto 5%;
}

.opinionCreate__cancelButon {
  margin: auto 5%;
  font-size: 1.125rem;
}
