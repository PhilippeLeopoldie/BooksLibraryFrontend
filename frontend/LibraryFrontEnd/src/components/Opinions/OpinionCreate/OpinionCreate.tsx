import { BookType, ReviewType } from "../../../constants/types";
import React, { useContext, useState } from "react";
import { OPINION_URL } from "../../../constants/api";
import { RateClick } from "../../Rates/RateClick/RateClick";
import "./OpinionCreate.css";
import { paginatedBooksCacheContext, ThemeContext, topBooksCacheContext } from "../../../App/App";

type AddOpinionType = {
    book?: BookType;
    toCreate: (value: string) => void;
    created: (value: boolean) => void;
};

export const OpinionCreate = ({ book, toCreate, created }: AddOpinionType) => {
    const theme = useContext(ThemeContext);
    const paginatedBooksCache = useContext(paginatedBooksCacheContext);
    const topBooksCache = useContext(topBooksCacheContext);
    const [bookCreatedMessage, setBookCreatedMessage] = useState<boolean>(false);
    const [errorOpinion, setErrorOpinion] = useState<boolean>(false);
    const [errorOpinionDetail, setErrorOpinionDetail] = useState<string>("");
    const [rateText, setRateText] = useState<string>("");
    const [formData, setFormData] = useState<ReviewType>({
        id:0,
        view: "",
        userName: "",
        rate: 0,
        postDate: "",
        bookId: 0
    });

    const RateTextConvertor = (rate: number) => {
        let rateText = "";
        switch (rate) {
            case 1:
                rateText = "Very bad";
                break;
            case 2:
                rateText = "Bad";
                break;
            case 3:
                rateText = "Good";
                break;
            case 4:
                rateText = "Very Good";
                break;
            case 5:
                rateText = "Excellent";
                break;
        }
        setRateText(rateText);
    };

    const HandleFormDataRate = (newRate: number) => {
        setFormData({ ...formData, rate: newRate });
        RateTextConvertor(newRate);
    };

    const PostOpinion = async (bookId: number) => {
        console.log("bookid", bookId);
        const requestOptions = {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                bookId,
                rate: formData.rate,
                view: formData.view,
                userName: formData.userName,
            }),
        };
        const opinionResponse: Response = await fetch(OPINION_URL, requestOptions);
        if (opinionResponse.status === 201) {
            const newOpinion = await opinionResponse.json();
            setBookCreatedMessage(true);
            setErrorOpinion(false);
            toCreate('bookPresentation');
            created(true);
            //newBooksCache?.setNewBooksCache([]);
            paginatedBooksCache?.setPaginatedBooks({
                paginatedItems: [],
                totalItems: 0,
                page: 0,
                totalPages: 0,
                requestedAt: ""
            });
            topBooksCache?.setTopBooksCache([]);
            return newOpinion;
        } else if (opinionResponse.status === 400) {
            const errorData = await opinionResponse.json();
            setBookCreatedMessage(false);
            setErrorOpinion(true);
            setErrorOpinionDetail(errorData.detail);
        }
    };

    const HideBookCreatedMessage = () => {
        setBookCreatedMessage(false);
    };
    const HideErrorDetail = () => {
        setErrorOpinion(false);
    };
    const HandleInputChange = (
        e:
            | React.ChangeEvent<HTMLInputElement>
            | React.ChangeEvent<HTMLTextAreaElement>
    ) => {
        setFormData({ ...formData, [e.target.name]: e.target.value });
        bookCreatedMessage && HideBookCreatedMessage();
        errorOpinion && HideErrorDetail();
    };

    return (
        <>
            <section
                className={`opinionCreate--${theme} opinionCreateCard--${theme}--grid opinionCreateCard--grid`}
            >
                <header className={`opinionCreate--${theme} opinionCreate__header`}>
                    <h2 className={`opinionCreate--${theme} opinionCreate__booktitle`}>
                        {book?.title}
                    </h2>
                    <h3 className={`opinionCreate--${theme} opinionCreate__bookauthor`}>
                        by: {book?.author}
                    </h3>
                </header>
                <main className={`opinionCreate--${theme} opinionCreate__main--grid`}>
                    <section className={`opinionCreate--${theme} opinionCreate__rate`}>
                        <RateClick rate={formData.rate} HandleRate={HandleFormDataRate} />
                        <div className={`opinionCreate--${theme} opinionCreate__rateText`}>
                            {rateText}
                        </div>
                    </section>
                    <section
                        className={`opinionCreate--${theme} opinionCreate__inputs--grid`}
                    >
                        <textarea
                            title="View"
                            className="opinionCreate__view"
                            placeholder="View"
                            name="view"
                            value={formData.view}
                            onChange={(e) => HandleInputChange(e)}
                        />
                        <input
                            className="opinionCreate__input"
                            placeholder="UserName"
                            name="userName"
                            value={formData.userName}
                            onChange={(e) => HandleInputChange(e)}
                        />
                        <section className={`opinionCreate--${theme} OpinionCreate_errorMessage`}>
                            {errorOpinion && (
                                <div
                                    className={`opinionCreate--${theme} validation__errorMessage`}
                                >
                                    {errorOpinionDetail}
                                </div>
                            )}
                        </section>
                    </section>
                </main>
                <footer className={`opinionCreate--${theme} opinionCreate__footer--flex`}>
                    <button
                        className="button opinionCreate__cancelButon"
                        onClick={() => {
                            toCreate('bookPresentation');
                        }}
                    >
                        Cancel
                    </button>
                    <button
                        className="button opinionCreate__postButton"
                        onClick={async () => {
                            book && book.id && (await PostOpinion(book.id));
                        }}
                    >
                        Post
                    </button>
                </footer>
            </section>
        </>
    );
};
