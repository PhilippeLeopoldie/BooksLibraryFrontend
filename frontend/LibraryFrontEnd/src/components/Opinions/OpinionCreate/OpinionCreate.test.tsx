import {render, screen,} from "@testing-library/react";
import { OpinionCreate } from "./OpinionCreate";
import "@testing-library/jest-dom";

describe("OpinionCreate", () => {
  test("snapShot: render Opinioncreate", ()=> {
    const {asFragment} = render(<OpinionCreate toCreate={()=> {}} created={()=>{}}/>)

    expect(asFragment()).toMatchSnapshot();
  })
})