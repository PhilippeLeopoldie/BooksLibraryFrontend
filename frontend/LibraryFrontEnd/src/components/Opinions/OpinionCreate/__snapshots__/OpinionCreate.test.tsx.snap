// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpinionCreate snapShot: render Opinioncreate 1`] = `
<DocumentFragment>
  <div
    class="opinionCreate--natural opinionCreateCard--natural--grid opinionCreateCard--grid"
  >
    <header
      class="opinionCreate--natural opinionCreate__header"
    >
      <h2
        class="opinionCreate--natural opinionCreate__booktitle"
      />
      <h3
        class="opinionCreate--natural opinionCreate__bookauthor"
      >
        by: 
      </h3>
    </header>
    <main
      class="opinionCreate--natural opinionCreate__main--grid"
    >
      <div
        class="opinionCreate--natural opinionCreate__rate"
      >
        <div
          class="rateClick_container--flex"
        >
          <div
            class="rate_star--natural rateClick_star empty"
          >
            ★
          </div>
          <div
            class="rate_star--natural rateClick_star empty"
          >
            ★
          </div>
          <div
            class="rate_star--natural rateClick_star empty"
          >
            ★
          </div>
          <div
            class="rate_star--natural rateClick_star empty"
          >
            ★
          </div>
          <div
            class="rate_star--natural rateClick_star empty"
          >
            ★
          </div>
        </div>
        <div
          class="opinionCreate--natural opinionCreate__rateText"
        />
      </div>
      <div
        class="opinionCreate--natural opinionCreate__inputs--grid"
      >
        <textarea
          class="opinionCreate__view"
          name="view"
          placeholder="View"
          title="View"
        />
        <input
          class="opinionCreate__input"
          name="userName"
          placeholder="UserName"
          value=""
        />
        <div
          class="opinionCreate--natural"
        />
      </div>
    </main>
    <footer
      class="opinionCreate--natural opinionCreate__footer--flex"
    >
      <button
        class="button opinionCreate__cancelButon"
      >
        Cancel
      </button>
      <button
        class="button opinionCreate__postButton"
      >
        Post
      </button>
    </footer>
  </div>
</DocumentFragment>
`;
